File: tools/conformance/src/validators/typescript/abstract-syntax-tree.ts
import { type Node, type SourceFile, SyntaxKind } from "typescript";

import {
  filterBySameKey,
  filterBySameKind,
  getChildren,
  getKey,
} from "./nodes";

import type { ConformanceError, ConformanceErrorLanguage } from "./types";

/**
 * Resolved error location.
 */
interface ResolvedErrorLocation {
  breadcrumb: string;
  instanceCharacter: number;
  instanceLine: number;
  snippet: string;
  templateCharacter: number;
  templateLine: number;
}

/**
 * Recursively walks the template and instance ASTs in lock-step, verifying
 * that every node present in the template also exists somewhere in the instance
 * at the same depth (superset semantics — the instance may contain extra nodes).
 *
 * For keyed template children the instance is searched by identity key.
 * For keyless template children all same-kind instance candidates are evaluated
 * by recursing into each; the candidate with the fewest errors is chosen and
 * its errors are propagated.
 *
 * Recursion bottoms out wherever the template has no children, so empty
 * method bodies and array literals in the instance are never checked.
 */
export function validateDepthFirstSearch(args: {
  instanceFile: SourceFile;
  instanceNode: Node;
  language: ConformanceErrorLanguage;
  templateNode: Node;
}): ConformanceError[] {
  const { instanceFile, instanceNode, language, templateNode } = args;
  const instanceChildren = getChildren(instanceNode);
  const templateChildren = getChildren(templateNode);
  const errors: ConformanceError[] = [];
  for (const templateChild of templateChildren) {
    errors.push(
      ...validateTemplateChild({
        instanceChildren,
        instanceFile,
        instanceNode,
        language,
        templateChild,
      }),
    );
  }
  return errors;
}

/**
 * Build error.
 */
function buildError(args: {
  instanceFile: SourceFile;
  instanceNode: Node;
  language: ConformanceErrorLanguage;
  templateChild: Node;
}): ConformanceError {
  const { instanceFile, instanceNode, language, templateChild } = args;
  const {
    breadcrumb,
    instanceCharacter,
    instanceLine,
    snippet,
    templateCharacter,
    templateLine,
  } = resolveErrorLocations({ instanceFile, instanceNode, templateChild });
  return {
    errorType: "code",
    fix: `Add the missing ${breadcrumb} to the instance file. See the template for the expected structure.`,
    instanceColumn: instanceCharacter + 1,
    instanceLine: instanceLine + 1,
    language,
    message: `Missing ${breadcrumb}`,
    templateColumn: templateCharacter + 1,
    templateLine: templateLine + 1,
    ...(snippet.length > 0 ? { expected: snippet } : {}),
  };
}

/**
 * Builds a structured `ConformanceError` describing a template node that is
 * absent from the instance.
 *
 * Captures:
 * - The source location within the instance file (where the child was expected).
 * - The source location within the template file (where the missing node lives).
 * - The `SyntaxKind` name of the missing node and its identity key when present.
 * - A condensed snippet of the template node's text as the `expected` value.
 */
function resolveErrorLocations(args: {
  instanceFile: SourceFile;
  instanceNode: Node;
  templateChild: Node;
}): ResolvedErrorLocation {
  const { instanceFile, instanceNode, templateChild } = args;
  const templateFile = templateChild.getSourceFile();
  const instancePos = instanceFile.getLineAndCharacterOfPosition(
    instanceNode.getStart(instanceFile),
  );
  const templatePos = templateFile.getLineAndCharacterOfPosition(
    templateChild.getStart(),
  );
  const kind =
    (SyntaxKind[templateChild.kind] as string | undefined) ??
    `SyntaxKind(${String(templateChild.kind)})`;
  const key = getKey(templateChild);
  return {
    breadcrumb: key === null ? kind : `${kind} "${key}"`,
    instanceCharacter: instancePos.character,
    instanceLine: instancePos.line,
    snippet: templateChild
      .getText(templateFile)
      .replaceAll(/\s+/gu, " ")
      .trim(),
    templateCharacter: templatePos.character,
    templateLine: templatePos.line,
  };
}

/**
 * Validate keyed child.
 */
function validateKeyedChild(args: {
  instanceChildren: Node[];
  instanceFile: SourceFile;
  instanceNode: Node;
  language: ConformanceErrorLanguage;
  templateChild: Node;
}): ConformanceError[] {
  const {
    instanceChildren,
    instanceFile,
    instanceNode,
    language,
    templateChild,
  } = args;

  const match = filterBySameKey(instanceChildren, templateChild)[0] ?? null;
  if (match === null) {
    return [
      buildError({ instanceFile, instanceNode, language, templateChild }),
    ];
  }

  return validateDepthFirstSearch({
    instanceFile,
    instanceNode: match,
    language,
    templateNode: templateChild,
  });
}

/**
 * Validate keyless child.
 */
function validateKeylessChild(args: {
  instanceChildren: Node[];
  instanceFile: SourceFile;
  instanceNode: Node;
  language: ConformanceErrorLanguage;
  templateChild: Node;
}): ConformanceError[] {
  const sameKind = filterBySameKind(args.instanceChildren, args.templateChild);

  if (sameKind.length === 0) {
    return [buildError(args)];
  }

  return sameKind
    .map((child) =>
      validateDepthFirstSearch({
        instanceFile: args.instanceFile,
        instanceNode: child,
        language: args.language,
        templateNode: args.templateChild,
      }),
    )
    .reduce((minimumErrors, currentErrors) =>
      currentErrors.length < minimumErrors.length
        ? currentErrors
        : minimumErrors,
    );
}

/**
 * Validate template child.
 */
function validateTemplateChild(args: {
  instanceChildren: Node[];
  instanceFile: SourceFile;
  instanceNode: Node;
  language: ConformanceErrorLanguage;
  templateChild: Node;
}): ConformanceError[] {
  const {
    instanceChildren,
    instanceFile,
    instanceNode,
    language,
    templateChild,
  } = args;
  if (getKey(templateChild) !== null) {
    return validateKeyedChild({
      instanceChildren,
      instanceFile,
      instanceNode,
      language,
      templateChild,
    });
  }
  return validateKeylessChild({
    instanceChildren,
    instanceFile,
    instanceNode,
    language,
    templateChild,
  });
}

File: tools/conformance/src/validators/python/abstract_syntax_tree.py
"""🌲 Depth-first search abstract syntax tree conformance validation."""

from typing import TYPE_CHECKING

from src.validators.python.nodes import (
    filter_by_same_key,
    filter_by_same_type,
    get_children,
    get_key,
)
from src.validators.python.types import ConformanceError

if TYPE_CHECKING:
    import ast


def _get_node_location(node: ast.AST) -> tuple[int | None, int | None]:
    line = getattr(node, "lineno", None)
    col = getattr(node, "col_offset", None)
    return line, (col + 1 if col is not None else None)


def _build_error(
    template_child: ast.AST, instance_node: ast.AST, filename: str
) -> ConformanceError:
    kind = type(template_child).__name__
    key = get_key(template_child)
    breadcrumb = f'{kind} "{key}"' if key is not None else kind
    instance_line, instance_col = _get_node_location(instance_node)
    template_line, template_col = _get_node_location(template_child)
    return ConformanceError(
        error_type="code",
        language="python",
        message=f"Missing {breadcrumb}",
        instance_line=instance_line,
        instance_column=instance_col,
        template_line=template_line,
        template_column=template_col,
        fix=f"Add the missing {breadcrumb} to the instance file. See the template for the expected structure.",
    )


def validate_depth_first_search(
    template_node: ast.AST,
    instance_node: ast.AST,
    template_source: str,
    instance_source: str,
    filename: str,
) -> list[ConformanceError]:
    errors: list[ConformanceError] = []
    instance_children = get_children(instance_node)
    template_children = get_children(template_node)
    for template_child in template_children:
        key = get_key(template_child)
        if key is not None:
            matches = filter_by_same_key(instance_children, template_child)
            match = matches[0] if matches else None
            if match is None:
                errors.append(_build_error(template_child, instance_node, filename))
            else:
                errors.extend(
                    validate_depth_first_search(
                        template_child, match, template_source, instance_source, filename
                    )
                )
        else:
            same_type = filter_by_same_type(instance_children, template_child)
            if not same_type:
                errors.append(_build_error(template_child, instance_node, filename))
            else:
                candidate_errors = [
                    validate_depth_first_search(
                        template_child, candidate, template_source, instance_source, filename
                    )
                    for candidate in same_type
                ]
                fewest = min(candidate_errors, key=len)
                errors.extend(fewest)
    return errors
